src/librustc_data_structures/sync.rs

```diff
@@ -497,13 +497,13 @@ impl<T> Once<T> {
     /// If the value was already initialized the closure is not called and `false` is returned,
     /// otherwise if the value from the closure initializes the inner value, `true` is returned
     #[inline]
-    pub fn init_locking<F: FnOnce() -> T>(&self, f: F) -> bool {
+    pub fn init_locking<F: FnOnce() -> T>(&self, f: F) -> &T {
         let mut lock = self.0.lock();
-        if lock.is_some() {
-            return false;
+        if lock.is_none() {
+            *lock = Some(f());
         }
-        *lock = Some(f());
-        true
+
+        self.borrow()
```

Comment:
hm, I suspect this won't work? i.e., the lock appears to still be held -- we can probably move the rest of the body into `{ ... }` or so

Reply:
Ouch, fixing it.